`Discussion`: Reference artifacts within postings

canberkanar opened the pull request:
<!-- Thanks for contributing to Artemis! Before you submit your pull request, please make sure to check the following boxes by putting an x in the [ ] (don't: [x ], [ x], do: [x]) -->
<!-- If your pull request is not ready for review yet, create a draft pull request! -->

### Checklist
#### General
<!-- You only need to choose one of the first two check items: Generally, test on the test servers. -->
<!-- If it's only a small change, testing it locally is acceptable and you may remove the first checkmark. If you are unsure, please test on the test servers. -->
- [x] I tested **all** changes and their related features with **all** corresponding user types on a test server.
- [x] Language: I followed the [guidelines for inclusive, diversity-sensitive, and appreciative language](https://docs.artemis.ase.in.tum.de/dev/guidelines/language-guidelines/).
- [x] I chose a title conforming to the [naming conventions for pull requests](https://artemis-platform.readthedocs.io/en/latest/dev/guidelines/development-process.html#naming-conventions-for-github-pull-requests).

#### Client
- [x] I followed the [coding and design guidelines](https://docs.artemis.ase.in.tum.de/dev/guidelines/client/) and ensured that the layout is responsive.
- [x] Following the [theming guidelines](https://docs.artemis.ase.in.tum.de/dev/guidelines/client-design/), I specified colors only in the theming variable files and checked that the changes look consistent in both the light and the dark theme.
- [x] I added multiple integration tests (Jest) related to the features (with a high test coverage), while following the [test guidelines](https://docs.artemis.ase.in.tum.de/dev/guidelines/client-tests/).
- [x] I documented the TypeScript code using JSDoc style.
- [x] I added multiple screenshots/screencasts of my UI changes.
- [x] I translated all newly inserted strings into English and German.

### Motivation and Context
In the current context, users refer to artifacts of Artemis such as lectures, attachments, and exercises byword in their postings. Due to this, others who want to get an idea of the mentioned artifact within a discussion first have to manually find and navigate to the artifact to access it. 

### Description
To enhance the user experience, we introduce two dropdowns to the markup editor of postings, 'Exercises' and 'Lectures'. When clicked, the list of the Exercises and Lectures of the current course is displayed. Attachments are also displayed in a nested way, where each is located under their specific Lecture. When a user selects an artifact, the corresponding markdown code is automatically generated and inserted into the markdown editor. Later when in preview mode or the posting is just viewed, the attachment's name is displayed as a link, together with an icon with respect to the type of the attachment.

### Steps for Testing
<!-- Please describe in detail how the reviewer can test your changes. -->
Prerequisites:
- 1 Student
- 1 Course with postings enabled

1. Log in to Artemis
2. Navigate to Course Discussions page
3. Create/edit postings with references to types (check the screencast for help)
- Post (can reference with : #{POST_ID} )
- File-upload Exercise
- Programming Exercise
- Modeling Exercise
- Text Exercise
- Quiz Exercise
- Lecture
- Lecture Attachment
4.  Verify that unique icons are displayed for each type of artifact referenced and make sure that they are consistent with the icons used in Artemis for those types.
5. Verify that previewing within the markdown editor works as well.
6. Make sure that all types of references are accessed when clicked.

### Review Progress
<!-- Each Pull Request should be reviewed by at least two other developers. The code as well as the functionality (= manual test) needs to be reviewed. -->
<!-- The reviewer or author check the following boxes depending on what was reviewed or tested. All boxes should be checked before merge. -->
<!-- You can add additional checkboxes if it makes sense to only review parts of the code or functionality. -->
<!-- When changes are pushed, uncheck the affected boxes. (Not all changes require full re-reviews.) -->

#### Code Review
- [x] Review 1
- [x] Review 2
#### Manual Tests
- [x] Test 1
- [x] Test 2

### Test Coverage
<!-- Please add the test coverages for all changed files here. You can see this when executing the tests locally (see build.gradle and package.json) or when looking into the corresponding Bamboo build plan. -->
<!-- Lines are the main reference but a significantly lower branch percentage can indicate missing edge cases in the tests. -->
| Class/File | Branch | Line |
|------------|-------:|-----:|
  exerciseReferenceCommand.ts | 100% | 100% |                                                                
  lectureAttachmentReferenceCommand.ts | 90% | 100% |
  metis.service.ts | 78.7% |  92.54% |
  metis.util.ts | 100% | 100% |                                                          
  posting-content.components.ts | 82.45% | 97.95% |                                    
  posting-content-part.components.ts | 91.66% | 100% | 
  posting-markdown-editor.component.ts | 77.77% | 92.85% |                                    
  negated-type-check.pipe.ts | 100% | 100% |                                                          



<!--
| ExerciseService.java | 85% | 77% |
| programming-exercise.component.ts | 13% | 95% |
-->

### Screenshots (UPDATED!!!)
<!-- Add screenshots to demonstrate the changes in the UI. -->
<!-- Create a GIF file from a screen recording in a docker container https://toub.es/2017/09/11/high-quality-gif-with-ffmpeg-and-docker/ -->
https://user-images.githubusercontent.com/67344158/174290816-0e3c5777-36ab-417e-aac6-b94ff07af3bb.mp4

jbllhff reviewed (dismissed):
Tested on ts2, works great, thanks!

akesfeden reviewed (changes requested):
Tested on TS1, references have correct icons and navigation works as expected, however Preview tab does not remove the closing tags when displaying them. I added screenshots below.

![image](https://user-images.githubusercontent.com/18427209/173679610-c470a6fd-0568-4cc4-af9f-49d4fbabbb62.png)

![image](https://user-images.githubusercontent.com/18427209/173679637-f4d4ca5b-7f82-47a7-af4b-e72d3c62801e.png)

TheTrueVester reviewed (dismissed):
Tested on ts1. Everything worked as expected.

Dominik-Weinzierl commented on src/main/webapp/app/shared/metis/posting-content/posting-content.components.ts:
```diff
@@ -56,14 +56,33 @@ export class PostingContentComponent implements OnInit, OnDestroy {
+                    ReferenceType.TEXT === referenceType ||
+                    ReferenceType.FILE_UPLOAD === referenceType
+                ) {
+                    referenceStr = this.content!.substring(this.content?.indexOf(')', patternMatch.startIndex)! + 1, patternMatch.endIndex - (referenceType.length + 3));
```
Could you please add some comments and an example to understand the string magic? :D e.g. why do you add 3, ...

Dominik-Weinzierl commented on src/test/javascript/spec/component/shared/metis/posting-content/posting-content-part.component.spec.ts:
```diff
@@ -1,30 +1,60 @@
                 component = fixture.componentInstance;
                 debugElement = fixture.debugElement;
+                router = TestBed.inject(Router);
+                metisService = TestBed.inject(MetisService);
```
![image](https://user-images.githubusercontent.com/60005702/173764354-b2569f18-c7e0-463c-b4dc-751ba6b3f68e.png)

Dominik-Weinzierl commented on src/test/javascript/spec/component/shared/metis/posting-content/posting-content-part.component.spec.ts:
```diff
@@ -68,9 +97,238 @@ describe('PostingContentPartComponent', () => {
+
+            // on click should navigate to referenced post within current tab
+            referenceLink.click();
+            expect(navigateByUrlSpy).toHaveBeenCalledTimes(1);
```
```suggestion
            expect(navigateByUrlSpy).toHaveBeenCalledOnce();
```

Dominik-Weinzierl commented on src/test/javascript/spec/component/shared/metis/posting-content/posting-content-part.component.spec.ts:
```diff
@@ -68,9 +97,238 @@ describe('PostingContentPartComponent', () => {
+            // on click should navigate to referenced post within current tab
+            referenceLink.click();
+            expect(navigateByUrlSpy).toHaveBeenCalledTimes(1);
+            expect(openAttachmentSpy).toBeCalledTimes(0);
```
```suggestion
            expect(openAttachmentSpy).not.toHaveBeenCalled();
```

Dominik-Weinzierl commented on src/test/javascript/spec/component/shared/metis/posting-content/posting-content-part.component.spec.ts:
```diff
@@ -68,9 +97,238 @@ describe('PostingContentPartComponent', () => {
+
+            // on click should navigate to referenced file upload exercise within current tab
+            referenceLink.click();
+            expect(navigateByUrlSpy).toHaveBeenCalledTimes(1);
+            expect(openAttachmentSpy).toBeCalledTimes(0);
```
again

Dominik-Weinzierl commented on src/test/javascript/spec/component/shared/metis/posting-content/posting-content-part.component.spec.ts:
```diff
@@ -68,9 +97,238 @@ describe('PostingContentPartComponent', () => {
+
+            // on click should navigate to referenced modeling exercise within current tab
+            referenceLink.click();
+            expect(navigateByUrlSpy).toHaveBeenCalledTimes(1);
+            expect(openAttachmentSpy).toBeCalledTimes(0);
```
again

Dominik-Weinzierl commented on src/test/javascript/spec/component/shared/metis/posting-content/posting-content-part.component.spec.ts:
```diff
@@ -68,9 +97,238 @@ describe('PostingContentPartComponent', () => {
+
+            // on click should navigate to referenced quiz exercise within current tab
+            referenceLink.click();
+            expect(navigateByUrlSpy).toHaveBeenCalledTimes(1);
+            expect(openAttachmentSpy).toBeCalledTimes(0);
```
again

Dominik-Weinzierl commented on src/test/javascript/spec/component/shared/metis/posting-content/posting-content-part.component.spec.ts:
```diff
@@ -68,9 +97,238 @@ describe('PostingContentPartComponent', () => {
+            expect(referenceLink).not.toBe(null);
+            expect(referenceLink.innerHTML).toInclude(referenceStr);
+
+            // should display relevant icon for lecture
+            const icon = getElement(debugElement, 'fa-icon');
+            expect(icon).not.toBe(null);
+            expect(icon.innerHTML).toInclude('fa fa-chalkboard-user');
+
+            // on click should navigate to referenced lecture within current tab
+            referenceLink.click();
+            expect(navigateByUrlSpy).toHaveBeenCalledTimes(1);
+            expect(openAttachmentSpy).toBeCalledTimes(0);
+        });
+
+        it('should contain a reference to attachment with icon', () => {
+            const referenceStr = 'Lecture 1 - Slide';
+            const attachmentURL = '/api/files/attachments/lecture/1/Lecture-1.pdf';
+            component.postingContentPart = {
+                contentBeforeReference,
+                referenceStr,
+                referenceType: ReferenceType.ATTACHMENT,
+                attachmentToReference: attachmentURL,
+                contentAfterReference,
+            } as PostingContentPart;
+            fixture.detectChanges();
+
+            // should display file name to user
+            const referenceLink = getElement(debugElement, '.reference');
+            expect(referenceLink).not.toBe(null);
+            expect(referenceLink.innerHTML).toInclude(referenceStr);
+
+            // should display relevant icon for attachment
+            const icon = getElement(debugElement, 'fa-icon');
+            expect(icon).not.toBe(null);
+            expect(icon.innerHTML).toInclude('fa fa-file');
+
+            // on click should open referenced attachment within new tab
+            referenceLink.click();
+            expect(openAttachmentSpy).toBeCalledTimes(1);
+            expect(openAttachmentSpy).toBeCalledWith(attachmentURL);
```
Please replace those tests with parameterized tests (this should be done really easy)

canberkanar replied:
Useful concept, thanks!

Dominik-Weinzierl reviewed (changes requested):
Code looks good - some minor suggestions

canberkanar commented on src/main/webapp/app/shared/metis/metis.service.ts:
```diff
@@ -356,6 +356,34 @@ export class MetisService implements OnDestroy {
+     * @param {string} exerciseId ID of the exercise to be navigated to
+     * @return {string} router link of the exercise
+     */
+    getLinkForExercise(exerciseId: string): string {
+        return `/courses/${this.getCourse().id}/exercises/${exerciseId}`;
+    }
+
+    /**
+     * returns the router link required for navigating to the lecture referenced within a posting
+     * @param {string} lectureId ID of the lecture to be navigated to
+     * @return {string} router link of the lecture
+     */
+    getLinkForLecture(lectureId: string): string {
+        return '/courses/' + this.getCourse().id + '/lectures/' + lectureId;
+    }
```
Good catch! I missed this one while refactoring code it seems.

TheTrueVester reviewed (dismissed):
Retested on ts1. Everything worked as expected.

Dominik-Weinzierl reviewed (approved):
Code changes looks good to me :+1: thank you for implementing my suggestions and for adapting the tests :+1:

alaraozdenler reviewed (approved):
Tested on ts2, everything works as described

ge65cer commented:
Great work! 👍 
Some suggestions: Would it be possible to also display the exercise icons in the dropdown list? Also, I'd be awesome if the dropdown is scrollable for courses with many exercises and lectures (currently, the whole modal scrolls up).

canberkanar commented:
> Great work! 👍 Some suggestions: Would it be possible to also display the exercise icons in the dropdown list? Also, I'd be awesome if the dropdown is scrollable for courses with many exercises and lectures (currently, the whole modal scrolls up).

Thank you for your review. Since the drop-down part is in the scope of markdown-editor and is designed in reusable way, the feature of supporting scrolling would be out of the scope of this PR. For instance, task and test cases dropdowns seen while editing the description of programming exercises are also allowed while using this feature. Adding support for icons is also not feasible in this shared component,.

minrows reviewed (dismissed):
Tested on ts2. LGTM 👍

akesfeden reviewed (dismissed):
Previous issues are solved, thank you for the fix! Working as expected and looks good.

Hialus reviewed (changes requested):
Tested on TS1 and I found some usability issues:
When creating a post for an exercise the model pops up and you have to scroll down the entire modal to select an exercise:
<img width="600" alt="grafik" src="https://user-images.githubusercontent.com/18218285/174076895-fb86fe8c-9010-4151-abad-b727652c7fcb.png">

<img width="600" alt="grafik" src="https://user-images.githubusercontent.com/18218285/174076852-97765e02-b1f0-42b3-9024-3e6367620e37.png">

This is even worse when it happens outside of a modal:
<img width="372" alt="grafik" src="https://user-images.githubusercontent.com/18218285/174077470-60f358be-1eb5-4796-8d44-cfedbf65a318.png">

You should consider making the dropdown content itself scrollable here.

Additionally, I don't like the usage of a simple dropdown for this. Especially if you have a lot of exercises (as seen in the screenshots above), it is very hard to find the exercise you are looking for. A search bar would help a lot here.

Another issue with this dropdown is its displaying on mobile devices. The dropdown can be cutoff as seen here:
<img width="400" alt="grafik" src="https://user-images.githubusercontent.com/18218285/174078739-783b8b48-f9c2-4c95-936c-79d13cf480eb.jpeg">
This could be fixed by dynamically aligning the content as explained here: [https://getbootstrap.com/docs/5.0/components/dropdowns/#responsive-alignment](https://getbootstrap.com/docs/5.0/components/dropdowns/#responsive-alignment) 

Minor issue: The `[programming]` tag is blue while the `[/programming]` is not 
<img width="540" alt="grafik" src="https://user-images.githubusercontent.com/18218285/174077292-36bd57d5-0bd9-4139-89f0-b25a4a9be24d.png">

While the last issue can be ignored, if it is not easy to fix, the first three cannot. They make this feature very user-unfriendly and should be fixed in this PR and not a follow-up.

Dominik-Weinzierl reviewed (changes requested):
I agree with @Hialus - you can provide a simple scrollable dropdown and add a search bar in a follow up for example (if this is too much for this PR)

canberkanar commented:
> Another issue with this dropdown is its displaying on mobile devices. The dropdown can be cutoff as seen here:

Thanks for your feedback.

- I added scrolling to the drop-down. I wasn't aware of the problem with mobile devices, yet it seems to me that the problem is solved there as well.
- The patterns I used to encode the references overlapped with Markup's own URL referencing command. That was the reason for the first section being in a different color than the trailer. Although I don't think this was really an issue, I changed the patterns so that some users are not distracted as in this case.
- It is up to you to approve this PR or not, but I am not adding a search function here. The exercise referencing feature was already merged and this "user-unfriendly" feature is already present in the code base. I will discuss this with Stephan when the time comes and will do the necessities in a follow-up PR later.

canberkanar commented:
> I agree with @Hialus - you can provide a simple scrollable dropdown and add a search bar in a follow up for example (if this is too much for this PR)

Thanks! I added scrolling to the dropdown menu.

Dominik-Weinzierl commented:
Thanks! Works as expected :+1: some suggestions and additional issues
- I think it's enough if you either highlight or underline the selected element. :thinking: 
![image](https://user-images.githubusercontent.com/60005702/174433089-cef48318-380c-4229-b3e3-cb723c783ee2.png)
- The lists (and quotes) do not work as expected
 ![image](https://user-images.githubusercontent.com/60005702/174433294-154cc537-0d90-4a40-bc7c-9e5d0bfb9ef9.png)
 ![image](https://user-images.githubusercontent.com/60005702/174433301-d6a04681-991c-4c71-80a9-ab8a27ffd812.png)

nadinesamyy reviewed (dismissed):
Works as expected!

Hialus commented:
> The exercise referencing feature was already merged and this "user-unfriendly" feature is already present in the code base. 

Your PR description says otherwise, which is why I remarked on it.
> To enhance the user experience, we introduce two dropdowns to the markup editor of postings, 'Exercises' and 'Lectures'. 

But the same issue would also apply to the lecture dropdown, as e.g. EIST also has a lot of lecture attachments, which would, again, make it hard for a user to find the one they want. This is by definition user-unfriendly. If I were to write a post referencing an exercise or lecture I would continue to just write the name of it instead of using this feature, as it would be quite a bit faster and simpler.
There are a couple of ways to make this feature more user-friendly:
- Sort the list. Currently, it appears to be either unsorted or sorted by ID.
- Provide a search bar in the dropdown
- Provide a way to add references while you are typing (with an auto-completion feature like used for test case names in Artemis or for PR/issue references on GitHub)
- Replace the dropdown with something else. E.g:
    - A modal
    - A sidebar (maybe with drag and drop functionality)
- Automatically add a reference when a user posts a link to an exercise/lecture/attachment

I still think that you should address this already in this PR and only create a follow-up PR to further improve it. While it is unfortunate that the exercise referencing feature has already been merged the PR description currently describes it to at least be modified by this PR. Therefore changing this would not be out of scope currently.

canberkanar commented:
> > The exercise referencing feature was already merged and this "user-unfriendly" feature is already present in the code base.
> 
> Your PR description says otherwise, which is why I remarked on it.
> 
> > To enhance the user experience, we introduce two dropdowns to the markup editor of postings, 'Exercises' and 'Lectures'.
> 
> But the same issue would also apply to the lecture dropdown, as e.g. EIST also has a lot of lecture attachments, which would, again, make it hard for a user to find the one they want. This is by definition user-unfriendly. If I were to write a post referencing an exercise or lecture I would continue to just write the name of it instead of using this feature, as it would be quite a bit faster and simpler. There are a couple of ways to make this feature more user-friendly:
> 
> * Sort the list. Currently, it appears to be either unsorted or sorted by ID.
> * Provide a search bar in the dropdown
> * Provide a way to add references while you are typing (with an auto-completion feature like used for test case names in Artemis or for PR/issue references on GitHub)
> * Replace the dropdown with something else. E.g:
>   
>   * A modal
>   * A sidebar (maybe with drag and drop functionality)
> * Automatically add a reference when a user posts a link to an exercise/lecture/attachment
> 
> I still think that you should address this already in this PR and only create a follow-up PR to further improve it. While it is unfortunate that the exercise referencing feature has already been merged the PR description currently describes it to at least be modified by this PR. Therefore changing this would not be out of scope currently.

Yes as you have mentioned, the exercise referencing is modified in this PR. When developing, I try to be consistent with what has been already merged in Artemis. Take a look at the screenshot below. I wonder if the drop-down here is not user-friendly, how did a very similar drop-down exists in production. This is very that reviewers' opinions are relative.

![Untitled](https://user-images.githubusercontent.com/67344158/174445195-badb02ca-c8d3-495a-8b7c-6dbd36c257b3.png)

Hialus commented:
> This is very sad that reviewers' opinions are relative.

That's just human nature. Reviews are always subjective to some degree, and reviewers look for different things when testing a feature (E.g., some reviewers only test if the testing steps work as described). This is also why more than one manual test is required to merge a PR.

And to answer your question: Yes, I do think that dropdown is user-unfriendly as well. It would be faster for me to directly go to the exercise page, than to go to this page and then select the exercise I want.

canberkanar commented:
> > * Provide a way to add references while you are typing (with an auto-completion feature like used for test case names in Artemis or for PR/issue references on GitHub)

I will check if this is feasible. For test cases, it works very well, but we have to check if it would be performance-wise okay searching through all the lectures, exercises, and attachments with each keyboard click.

TheTrueVester reviewed (approved):
Retested on ts1. Everything worked as expected.

max-bergmann reviewed (approved):
Tested on TS1, works as expected

SimonPaul02 reviewed (approved):
Tested on TS 2, works as expected. But I guess the students won't know how to reference other posts. It would be nice have some kind of dropdown option for posts. But this is not part of this PR, good job!

yuling-sun reviewed (approved):
Tested on TS2. References work as expected. It's not intuitive that references to other posts are done with #{POST_ID}.

canberkanar commented:
> Tested on TS 2, works as expected. But I guess the students won't know how to reference other posts. It would be nice have some kind of dropdown option for posts. But this is not part of this PR, good job!

Thanks! Referencing posts was part of a previous development, but we will consider your opinion for future improvements. 👍🏻

ole-ve commented:
Nice feature.
Regarding the lecture references, I'd have two comments:
<img width="138" alt="Bildschirmfoto 2022-06-21 um 11 10 53" src="https://user-images.githubusercontent.com/73833780/174763070-2595008e-c7f5-4c56-b0f3-57cd7c28aa8f.png">

1. To me as a user, it is not clear which entry of `Lecture 2` to choose. Did you simply create multiple lectures with the same name or multiple lecture units with the same name? In addition, a small icon for every entry that displays if the entry represents a lecture/attachment may also help.
2. Agreeing to Timor, a dropdown can become user-unfriendly very fast. Especially, when there are a lot of lectures/attachments (in EIST: 14 Lectures á 5 Slidesets = 70 Attachments). However, using a modal makes the process of adding references more complex and may not be used as a result. Therefore, I'd suggest adding a search bar to query for units.

babyygemperor reviewed (approved):
Tested on ts2, works as described

canberkanar commented:
> Thanks! Works as expected 👍 some suggestions and additional issues
> 
> * I think it's enough if you either highlight or underline the selected element. 🤔
>   ![image](https://user-images.githubusercontent.com/60005702/174433089-cef48318-380c-4229-b3e3-cb723c783ee2.png)
> * The lists (and quotes) do not work as expected
>   ![image](https://user-images.githubusercontent.com/60005702/174433294-154cc537-0d90-4a40-bc7c-9e5d0bfb9ef9.png)
>   ![image](https://user-images.githubusercontent.com/60005702/174433301-d6a04681-991c-4c71-80a9-ab8a27ffd812.png)

Thanks for the feedback once again.

1.  I have removed the underlining on hover.
2. The problem is not in scope of this PR as the same problem can be replicated by referencing a post within an ordered list as well. Here is the issue I have created for your reference, which will be fixed in the future. #5281 We discussed the issue with Stephan today and decided that this PR can be merged.

canberkanar commented:
> Nice feature. Regarding the lecture references, I'd have two comments: <img alt="Bildschirmfoto 2022-06-21 um 11 10 53" width="138" src="https://user-images.githubusercontent.com/73833780/174763070-2595008e-c7f5-4c56-b0f3-57cd7c28aa8f.png">
> 
> 1. To me as a user, it is not clear which entry of `Lecture 2` to choose. Did you simply create multiple lectures with the same name or multiple lecture units with the same name? In addition, a small icon for every entry that displays if the entry represents a lecture/attachment may also help.
> 2. Agreeing to Timor, a dropdown can become user-unfriendly very fast. Especially, when there are a lot of lectures/attachments (in EIST: 14 Lectures á 5 Slidesets = 70 Attachments). However, using a modal makes the process of adding references more complex and may not be used as a result. Therefore, I'd suggest adding a search bar to query for units.

1. Multiple lectures with the same name were created for testing purposes.
2. We will consider your suggestion for future improvements. Thanks!
